lib/Parse/ParseExpr.cpp

```diff
@@ -2928,7 +3034,15 @@ bool Parser::ConsumeAndStoreBoundsExpression(CachedTokens &Toks) {
     return false;
   }
   ConsumeParen();
-  return ConsumeAndStoreUntil(tok::r_paren, Toks, /*StopAtSemi=*/true);
+  result = ConsumeAndStoreUntil(tok::r_paren, Toks, /*StopAtSemi=*/true);
+  if (Tok.getKind() == tok::identifier) {
+    Toks.push_back(Tok);
+    ConsumeToken();
+    Toks.push_back(Tok);
+    ConsumeParen();
```

Comment:
Please guard the `ConsumeParen` before calling it.  It should be guarded the same way the prior `ConsumeParen` call is guarded:
```
if (Tok.getKind() != tok::l_paren) {
  return false;
}
```
ConsumeParen emits an error message if it is called when the token is not a parentheses.   We are storing tokens now to be parsed later, so we don't want error messages here.